controllers/topology/internal/extensions/patches/inline/json_patch_generator.go

```diff
@@ -0,0 +1,293 @@
+		return nil, errors.Errorf("failed to calculate value: .valueFrom is set, but both .valueFrom.variable and .valueFrom.template are set")
+	}
+
+	// Return raw value.
+	if patch.Value != nil {
+		return patch.Value, nil
+	}
+
+	// Return variable.
+	if patch.ValueFrom.Variable != nil {
+		value, ok := variables[*patch.ValueFrom.Variable]
+		if !ok {
+			return nil, errors.Errorf("failed to calculate value: variable %q does not exist", *patch.ValueFrom.Variable)
+		}
+		return &value, nil
+	}
+
+	// Return rendered value template.
+	value, err := renderValueTemplate(*patch.ValueFrom.Template, variables)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to calculate value")
+	}
+	return value, nil
+}
+
+// renderValueTemplate renders a template with the given variables as data.
+func renderValueTemplate(valueTemplate string, variables map[string]apiextensionsv1.JSON) (*apiextensionsv1.JSON, error) {
+	// Parse the template.
+	tpl, err := template.New("tpl").Parse(valueTemplate)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to parse template: %q", valueTemplate)
+	}
+
+	// Convert the flat variables map in a nested map, so that variables can be
+	// consumed in templates like this: `{{ .builtin.cluster.name }}`
+	// NOTE: Variable values are also converted to their Go types as
+	// they cannot be directly consumed as byte arrays.
+	data, err := calculateTemplateData(variables)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to calculate template data")
+	}
+
+	// Render the template.
+	var buf bytes.Buffer
+	if err := tpl.Execute(&buf, data); err != nil {
+		return nil, errors.Wrapf(err, "failed to render template: %q", valueTemplate)
+	}
+
+	// Unmarshal the rendered template.
+	// NOTE: The YAML library is used for unmarshalling, to be able to handle YAML and JSON.
+	value := apiextensionsv1.JSON{}
+	if err := yaml.Unmarshal(buf.Bytes(), &value); err != nil {
+		return nil, errors.Wrapf(err, "failed to unmarshal rendered template: %q", buf.String())
+	}
+
+	return &value, nil
+}
+
+// calculateTemplateData calculates data for the template, by converting
+// a flat variables map to a nested map and converting the variables to
+// their Go types.
+// Example:
+// * Input:
+//   map[string]apiextensionsv1.JSON{
+//     "builtin.cluster.name": {Raw: []byte(`"cluster-name"`},
+//     "integerVariable": {Raw: []byte("4")},
+//     "numberVariable": {Raw: []byte("2.5")},
+//     "booleanVariable": {Raw: []byte("true")},
+//   }
+// * Output:
+//   map[string]interface{}{
+//     "builtin": map[string]interface{}{
+//       "cluster": map[string]interface{}{
+//         "name": <string>"cluster-name"
+//       }
+//     },
+//     "integerVariable": <float64>4,
+//     "numberVariable": <float64>2.5,
+//     "booleanVariable": <bool>true,
+//   }
+func calculateTemplateData(variables map[string]apiextensionsv1.JSON) (map[string]interface{}, error) {
```

Comment:
Could we json.Marshal/Unmarshal into a map for easy conversion?

Reply:
Follow-up question: Are we disallowing variables to have dots inside?

Reply:
> Could we json.Marshal/Unmarshal into res for easy conversion instead of writing our own function?

I'm not sure how this could work, given that the builtin variables are nested to different depths and we have to unmarshal the values so that they can be used correctly in Go templates (and there are also user-defined variables which are not known up-front). Can you provide an example?


> I understand here that we're also expanding the variables, although why not having the builtin variables to be already as a nested map instead?

We currently only need the nested map to make the go templating syntax possible. 

I'm not sure that we want to have nested variables overall as a concept. Currently, we're only having a "flat" variable list wheres the variables themselves are currently basic types but will probably also be complex types in the future. (I'm assuming having nested builtin variables means that we're passing the nested maps to patch generators, i.e. also to external patching)

Reply:
We could do ~ the following:
```
	tmp, _ := json.Marshal(variables)
	json.Unmarshal(tmp, &res)
```
This converts the `apiextensionsv1.JSON` values to their go type (string, float64, ..), but we would still have to convert the map from flat to nested.

I'm also not sure if it's a good idea given that in other places we already starting to try to optimize performance so a full Marshal/Unmarshal over all variables might not be ideal compared to only unmarshalling the values.

Reply:
The latter is what I was suggesting

Reply:
As discussed, refactored so that builtin variables are now just **a** builtin variable (object)